V1.0

kaankaray opened the pull request:
App is fully functional. Please use Bearer Auth and send JSON queries. Postman config exported to commit.

Zakaria98-dz committed:
i made some changes in the handlers.py and routes.py and the storage.py where I added the user' email and password information
note that there is no logical function in it

kaankaray committed:
- Changed some method names.
- Removed key from pattern.
- Logic has [get,set]_user_count methods and an increase_user_count method.
- User creation returns key now.

kaankaray committed:
- User object is implemented.
- Create user, login and login functions fully work now according to the specification.
- Create user now has a few checks.

- Auth tokens are now kept in 'authTokens'
- Some User regarding methods to make things easier.
- Deprecated some old methods

- /user/create does not require key now.
- /user login method is implemented(Needs some checks).
- /user/logout is implemented(Needs some checks).

kaankaray committed:
User can create shortlinks, and save it on the both user and separately. More things to do, taking a break

kaankaray committed:
Everything works.
- Authentication works with a Bearer
- Shorten url will work

kaankaray committed:
final descriptions

kaankaray committed:
Add files via upload

Postman queries exported.

anxolerd commented on url_shortener/logic.py:
```diff
@@ -12,15 +12,91 @@ def __init__(self, storage: Storage):
         """ Creating an instance of our logic. Pay attention that storage is not created within the logic, but
         passed as a parameter from outside. This is called Dependency Injection.
         """
+        print("Logic is initilized.")
```
If these `print` statements were added here for debugging purposes, they should be removed. If they are needed for infromational purposes on application startup, please use `logging` module:

```python
import logging

# The following line will instantiate a logger instance which knows about current module name
# so you will be able to know which module logged which lines
log = logging.getLogger(__name__)

# Simple log. Available functions are `debug`, `info`, `warn`, `error`.
log.info('Your message here')

# You can interpolate some variables into your log message:
answer = 42
log.info('Answer is %s. Unfortunately I do not know the question', answer)
```

anxolerd commented on postman.json:
```diff
@@ -0,0 +1,265 @@
+{
```
It is good, you included this file, so others can easily check out your app :+1:

anxolerd commented on url_shortener/dto.py:
```diff
@@ -4,6 +4,31 @@
+
+
+    def generate_token(self):
+        self.token = hash(self.username)
```
User should be able to login from multiple locations with different tokens, so that they could logout only from one device, but not from another (for example, user logged in from their mobile phone and laptop, than they decided to logout from the phone, but keep the laptop session active).

I would rather not store login token in `User` entity, but keep a separate dictionary which maps tokens to users.

Also, it would be better if you generate random tokens instead of relying on `hash` function. For example, you can use [UUID](https://docs.python.org/3/library/uuid.html) module for that:

```python
import uuid

token = uuid.uuid4().hex
```

kaankaray replied:
Thank you! Although let's assume that we have 5 logged in users, and let's assume that for some reason all of them logged in in 3 different devices. Iterating through logged in users(there is a list) and then inside of this user, finding this UUID would be faster. Then iterating through all 15 elements.

userObject.token should be an List<str> and keep UUIDs 

What do you think?

anxolerd replied:
> 5 logged in users all of them logged in in 3 different devices. Iterating through logged in users(there is a list) and then inside of this user, finding this UUID would be faster. Then iterating through all 15 elements.

In first cases you will iterate in worst case through 3 * 5 = 15 items. This will take almost the same time as iterating through 15 elements. However, there are data structures that allow constant-time search for the element. You can use `dict` structure for that:

```python
tokens_to_users = {
    'abcd-abcd-abcdabcdabcd-0001': 'user1',
    'abcd-abcd-abcdabcdabcd-0002': 'user2',
    'abcd-abcd-abcdabcdabcd-0003': 'user1',
    ...    
}
user_id = tokens_to_user.get(token)
```

This will find a `user_id` in constant time by token and that will be faster than iterating through all elements.

anxolerd commented on url_shortener/views/handlers.py:
```diff
@@ -1,60 +1,332 @@
+            'description': 'Request body is missing.'
+        })
+
+    # Conditions.
```
You can use [`trafaret`](https://trafaret.readthedocs.io/en/latest/intro.html) library included in the project for validation.

anxolerd commented on url_shortener/views/handlers.py:
```diff
@@ -1,60 +1,332 @@
+
+    # If custom id is not provided, or this id already exists, we give an random one.
+    while id_provided is None or logic.read_by_key(id_provided) is not None:
+        id_provided = ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase) for _ in range(4))
```
1. This looks like a part of business logic. Should it really be in the handler, or should it be in the `logic` module?
2. When you use all 4-character strings your application will stuck here in the infinite loop :(

anxolerd commented on url_shortener/views/handlers.py:
```diff
@@ -1,60 +1,332 @@
+    if any(c.isdigit() for c in id_provided):
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Numbers in id is not allowed.'
```
But why numbers are not allowed? What if I want to have shortened url like `https://localhost:6543/lab_1` ?

kaankaray replied:
We store users in key: value fashion. Did not wanted user to overwrite on users. 

Improved this by adding a condition of id should contain at least one alphabetic char.

anxolerd commented on url_shortener/views/handlers.py:
```diff
@@ -1,60 +1,332 @@
+            'status': 'success',
+            'description': 'Successfully created short url.',
+            'id': id_provided,
+            'shortened_link': domain_name + id_provided
```
Instead of defining `domain_name` constant here, you can read `request.registry.base_url`

anxolerd commented on url_shortener/views/handlers.py:
```diff
@@ -1,60 +1,332 @@
-            headerlist=[('Location',
-                         request.registry.base_url + '/resource/' + key)],
-        )
+        new_user = logic.find_user_by_token(get_auth_token(request))
```
It is my bad for not showing this in the example project, but you can get user email by accessing `request.authenticated_userid`. Reference: https://docs.pylonsproject.org/projects/pyramid/en/latest/api/request.html#pyramid.request.Request.authenticated_userid

anxolerd reviewed (changes requested):
Good job overall, but it lacks tests. Please add some unit-tests for your business logic.

anxolerd commented on url_shortener/views/handlers.py:
```diff
@@ -1,60 +1,332 @@
+        })
+
+    # URL check if it's valid. (stolen from django url validation regex)
+    regex_check = re.compile(
+        r'^(?:http|ftp)s?://'  # http:// or https://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
+        r'localhost|'  # localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
+        r'(?::\d+)?'  # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+
+    if not re.match(regex_check, url_provided.lower()) is not None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'URL is incorrect.'
+        })
```
Write tests for this, as this looks quite complicated.

kaankaray committed:
Change requests from pull req. 4.
- Now we don't save user, but the {UUID as str: username as str}
- Moved random text generator to logic.py
- Random text generator now generates with numbers.
- Random text generator generates with an extra character or digit if operation takes too long.
- Digits are allowed to be IDs now. As long as they have at least one char
- instead of writing domain_name manually we use request.registry.base_url now.

TODO: user logged in cannot be checked.
TOOD: unit tests!

kaankaray commented on url_shortener/views/handlers.py:
```diff
@@ -126,26 +126,33 @@ def login_user(request: Request) -> Response:
+    print(request.authorization) # Returns None as expected
+    print(request.authenticated_userid) # Returns None
+    print(token)
+    if request.is_authenticated:
```
I don't know how can I check if user is logged in through this device. UUID doesn't provide a HWID. And forgot to comment, if statement always returns False

kaankaray replied:
forgot to comment, if statement always returns False

kaankaray commented on url_shortener/views/__init__.py:
```diff
@@ -1,3 +1,5 @@
 
 __all__ = ['setup_routes']
+
+# TODO: unit tests! Dunno what to test really
```
Other than REGEX checks, I don't know what else to check

anxolerd replied:
Nothing to test in this file actually.


Start with writing tests for your regexes. Check whether strings that should match according to your business logic do match actually and strings that should not --- do not match.

Secondly you have some `conditions` or input validations based on length, etc. Tests whether it works according to your business rules and that the code does exactly what you wanted it to do.

Moreover, you have storage implementation with some searches and logic built in. Test those are working correctly. For example, test that when you search for links created by one user you do not receive links created by different user.

kaankaray reviewed (commented):
Overall tried to fix all the issues mentioned above. Most of them makes sense! Thank you! Stuck at 2 points though:

kaankaray committed:
- Wrong test method of email checking.
- is_url_valid check was returning (not) version of it

kaankaray committed:
Some tests, login and logout test are TODO

anxolerd reviewed (approved):
Great Job!

Please prepare a course work document describing what did you implement here and how did you do that =). It would be great if you prepare the course work before Jun 22, 2021.
